axum/src/routing/url_params.rs

```diff
@@ -0,0 +1,43 @@
+use crate::util::{ByteStr, PercentDecodedByteStr};
+use http::Extensions;
+use matchit::Params;
+
+pub(crate) enum UrlParams {
+    Params(Vec<(ByteStr, PercentDecodedByteStr)>),
+    InvalidUtf8InPathParam { key: ByteStr },
+}
+
+pub(super) fn insert_url_params(extensions: &mut Extensions, params: Params) {
+    if let Some(UrlParams::InvalidUtf8InPathParam { .. }) = extensions.get() {
+        // nothing to do here since an error was stored earlier
+        return;
+    }
+
+    let params = params
+        .iter()
+        .filter(|(key, _)| !key.starts_with(super::NEST_TAIL_PARAM))
+        .map(|(key, value)| (key.to_owned(), value.to_owned()))
+        .map(|(k, v)| {
+            if let Some(decoded) = PercentDecodedByteStr::new(v) {
+                Ok((ByteStr::new(k), decoded))
+            } else {
+                Err(ByteStr::new(k))
+            }
+        })
+        .collect::<Result<Vec<_>, _>>();
+
+    match (extensions.get_mut(), params) {
```

Comment:
Is there any reason to look up the extension a second time here instead of just beginning the function with `let extension = extensions.get_mut();` and basing the short-circuiting on that?

Reply:
Nope. I just missed it